### ISSUE TITLE
Parser for generic ASCII tables.

### ISSUE THREAD

**Issue by CtrlC-Root** [kellyjonbrazil/jc#158]
I have a few tools I use that output their data as ASCII tables in various formats. An example would be the `virsh` commnad line interface:

```
$ virsh pool-list --all                                                                                                                                            
 Name          State    Autostart
-----------------------------------
 default       active   no
 disk-images   active   yes
 vmdisk        active   yes
```

The issue is that `virsh` alone has hundreds of commands which output different columns. Some of those commands even allow selectively hiding or including specific columns in the output.

I'd like to parse the output of these but I don't know that it makes sense to write a parser for each one or even for each of these programs. What if we added a generic ASCII table parser instead that could handle some common output formats? As an example here's a short Python 3 script I wrote as a first pass at this: https://github.com/CtrlC-Root/dotfiles/blob/master/bin/tj. Would something like this make sense to add to this project?

**Comment by kellyjonbrazil** [comment]
Interesting - I do have a `universal` parser that works on columnar tables output like `ps` and `netstat`. It's an internal parser used by the higher-level parsers:

https://github.com/kellyjonbrazil/jc/blob/master/jc/parsers/universal.py

A higher-level ASCII Table parser could probably be built on top of one or both of the Universal parsers.

**Comment by kellyjonbrazil** [comment]
Hi @CtrlC-Root any thoughts on bringing this back to life? I think this would be a cool parser! If you don't have plans to pick this up, maybe I'll take a stab at it.

Thanks!

**Comment by kellyjonbrazil** [comment]
I have a couple of ASCII table parsers working in `dev` now.

The first is the standard, single line table parser and it can handle all sorts of ASCII and Unicode tables:
```
$ echo '                                                                                                                                           
>  Name          State    Autostart
> -----------------------------------
>  default       active   no
>  disk-images   active   yes
>  vmdisk        active   yes' | jc --asciitable -p
[
  {
    "name": "default",
    "state": "active",
    "autostart": "no"
  },
  {
    "name": "disk-images",
    "state": "active",
    "autostart": "yes"
  },
  {
    "name": "vmdisk",
    "state": "active",
    "autostart": "yes"
  }
]
```
Markdown tables:
```
$ echo '
>     | type   |   total |   used |    free |   shared |   buff cache |   available |
>     |--------|---------|--------|---------|----------|--------------|-------------|
>     | Mem    | 3861332 | 222820 | 3364176 |    11832 |       274336 |     3389588 |
>     | Swap   | 2097148 |      0 | 2097148 |          |              |             |' | jc --asciitable -p
[
  {
    "type": "Mem",
    "total": "3861332",
    "used": "222820",
    "free": "3364176",
    "shared": "11832",
    "buff_cache": "274336",
    "available": "3389588"
  },
  {
    "type": "Swap",
    "total": "2097148",
    "used": "0",
    "free": "2097148",
    "shared": null,
    "buff_cache": null,
    "available": null
  }
]
```
Fancy Unicode tables:
```
$ echo '
>     ╒══════════╤═════════╤════════╕
>     │ foo      │ bar     │ baz    │
>     ╞══════════╪═════════╪════════╡
>     │ good day │         │ 12345  │
>     ├──────────┼─────────┼────────┤
>     │ hi there │ abc def │ 3.14   │
>     ╘══════════╧═════════╧════════╛' | jc --asciitable -p
[
  {
    "foo": "good day",
    "bar": null,
    "baz": "12345"
  },
  {
    "foo": "hi there",
    "bar": "abc def",
    "baz": "3.14"
  }
]
```
You can even parse fancy tables with multi-line rows with the `asciitable-m` parser:
```
$ echo '
>     ╒══════════╤═════════╤════════╕
>     │ foo      │ bar baz │ fiz    │
>     │          │         │ buz    │
>     ╞══════════╪═════════╪════════╡
>     │ good day │ 12345   │        │
>     │ mate     │         │        │
>     ├──────────┼─────────┼────────┤
>     │ hi there │ abc def │ 3.14   │
>     │          │         │        │
>     ╘══════════╧═════════╧════════╛' | jc --asciitable-m -p
[
  {
    "foo": "good day\nmate",
    "bar_baz": "12345",
    "fiz_buz": null
  },
  {
    "foo": "hi there",
    "bar_baz": "abc def",
    "fiz_buz": "3.14"
  }
]
```

**Comment by CtrlC-Root** [comment]
This is fantastic! I've wanted to pick this back up but I just haven't found the time to even respond here much less work on it. I have a script that parses virsh output well enough and I've just been using that but I look forward to dropping it once this is implemented.

**Comment by kellyjonbrazil** [comment]
@CtrlC-Root Thanks for the inspiration on this one. It's a cool idea!

Just an FYI - if you do want to test before the next jc release you can copy the parser files to your [plugin parser directory](https://github.com/kellyjonbrazil/jc#custom-parsers)

Here are the parser files:
https://github.com/kellyjonbrazil/jc/blob/dev/jc/parsers/asciitable.py
https://github.com/kellyjonbrazil/jc/blob/dev/jc/parsers/asciitable_m.py

I'll probably add these to v1.18.6 and release in a week or so.

**Comment by kellyjonbrazil** [comment]
`asciitable` and `asciitable-m` parsers have been released with jc v1.18.6
